### PR TITLE
feat(graphql): allow the ability to ignore certain fields

### PR DESCRIPTION
**Pull request by jscherer92**
## Which problem is this PR solving?
The inability to specify fields that you may want to ignore due to high levels processing needed to get through all of those fields.

## Short description of the changes
Add a new option into the config to ignore fields.

## Checklist

- [x] Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable.

### CODE REVIEW

**Comment by jscherer92** [comment]
@dyladan sorry that this is so late and it does not have code coverage on the new code, but this is what I was talking about in our one meeting. When you have arrays that are 20,000 objects that each of those objects have 20-40 properties, it leads to a major slowdown. This is one way of fixing the issue since it would allow us to ignore fields that we do not want to put into the tracing.

It also should allow us to hide values from entering into our telemetry platform. If you think this is a good idea, I'll continue with this and add an actual feature and finish this up.

**Comment by jscherer92** [comment]
@Flarna @blumamir or others, can I get a set of eyes on this? Will help me do some more things with this instrumentation from a PII perspective

**Comment by Flarna** [comment]
@jscherer92 I'm not familiar with graphql so can't really comment that much on this change.
The current owner of this instrumenation is @obecny but as far as I know he moved to other topics then OTel a while ago.

@open-telemetry/javascript-maintainers  what is the right approach here? I think we should be happy to have people to contribute but the contribution guideline is not that clear to me how to move ownership.

File: plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts:303]: when you instantiate class you could create a hash map of all keys to ingore using `includes` everywhere which is slow.

**Review by obecny — COMMENTED** [review]
please improve the performance - create hash map at start from config array property

**Comment by jscherer92** [comment]
Would a HashSet be better?

**Comment by obecny** [comment]
> Would a HashSet be better?

Pseude code

```
this._ignorehMap = {};
for (let i = 0, j = ignoreKeys.length; i < j; i++) {
  this._ignorehMap[ignoreKeys[i]] = true;
}


if (
    !type ||
    typeof type.getFields !== 'function' ||
    this._ignorehMap(type.name) ||
...
```

**Comment by jscherer92** [comment]
Oh I agree that we could do it that way. I'm just wondering if the HashSet would be better;

`
this.ignoreFields = new HashSet(this.config.ignoreFields);


if(this.ignoreFields.has(field))
`

**Comment by obecny** [comment]
> Oh I agree that we could do it that way. I'm just wondering if the HashSet would be better;
> 
> ` this.ignoreFields = new HashSet(this.config.ignoreFields);
> 
> if(this.ignoreFields.has(field)) `

hash map is faster the set

**Comment by jscherer92** [comment]
should be switched over now

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:78]: map is meant to return a new array if you want to iterate please use forEach instead

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:82]: this seems to be over engineering , a private function that returns a private, I think you can get rid of this, unless there is very specific reason for that ?

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:87]: this whole function can be removed

File: plugins/node/opentelemetry-instrumentation-graphql/src/types.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-graphql/src/types.ts:60]:
```suggestion
   * List of fields name / key  to be ignored for span creation.
```

File: plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts:372]: you created a map but you still didn't use it

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:78]: you already created a new map, then you create a new new one and then you created a new private function to override the created at the beginning. 
why cant you simply do
```
config.ignoreFields?.forEach(item => this._ignoreMap.set(item, true));
```
?

**Review by obecny — CHANGES_REQUESTED** [review]

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (jscherer92) [reply plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts]: It is now in two locations for the set behavior, so I kept this but repurposed

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (jscherer92) [reply plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts]: the reason is due to the wrapField and wrapFieldResolver functions. Since it can on the fly change the config, we need to make sure that those util functions get the updates. There are embedded functions so this will make sure we don't have to rerun that code.

File: plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts
Comment (jscherer92) [reply plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts]: fixed

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:67]: this is always overridden in  `_setIgnoreMap` so I think it should be safe to remove assigning this to new map here ?

**Review by obecny — APPROVED** [review]
There is one comment which I hope you can address before merging. Other than than lgtm, thx for changes.

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (jscherer92) [reply plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts]: Yep! You are definitely right, fixed it in the latest change

File: plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts
Comment (Rauno56) [inline plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts:303]: Is the indirection required?

Comment (Rauno56) [reply plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts:303]: OK, saw https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1134#discussion_r963077861 now.
The whole thing about config changing is such an edge case I think, but alright - I guess that's another topic.

It shouldn't be an issue if you never overwrite the variable. Map has a `clear()` method just for that.

**Comment by Rauno56** [comment]
> hash map is faster the set

I cannot understand how that can even be the case. I mean... even if it would, why didn't the smart people behind node think of implementing Set through Map then?

**Comment by Rauno56** [comment]
GraphQL tests seem to be flaky(see #1184). Merging this will be blocked until that's fixed.

@jscherer92, maybe you'd be willing to look into it?

**Comment by Rauno56** [comment]
Nvm, GraphQL tests are now fine.

File: plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts:376]: Can we use `info.fieldName` instead of `path[path.length - 1]`?
Seems to me more straightforward and robust, than using the output of the `pathToArray`

Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts:376]: What is the motivation of running the resolver with the context of parent span, instead of just not instrumenting this call?

If I understand correctly, then for this query
```
{
  foo {
    bar
  }
}
```

if bar is ignored, then downstream spans created under the `bar` resolver will be children of `foo`, right?

Is this what we want?

File: plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts:315]: If I understand the flow correctly, then this will not only ignore this field, but also all the schema resolver field below it, e.g.
```
{
  foo {
    bar
  }
}
```

Then ignoring only `foo` will terminate this recursion function, preventing it from calling `wrapFields` below on `foo`s field resolver, including `bar`.

This will apply for resolvers on the schema, but for "fieldResolver" invocations, the behavior will be different, as nothing will prevent it from being called on `bar` and thus the span will be recorded, right?

I suggest refactoring the code to have consistency between the 2 resolver options, and adding a test for it.

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:67]: @obecny - you wrote
> hash map is faster the set

I guess you refer to this issue?
https://stackoverflow.com/questions/49841972/javascript-how-come-map-has-is-so-much-faster-than-set-has-and-array-indexof

I don't know the current status of the issue, [but from looking on node source code](https://github.com/nodejs/node/blob/36805e852471cc7a24524a329848da5c8f163864/deps/v8/src/compiler/js-call-reducer.cc#L7557), it seems that `Set.has` is nowdays treated the same as `Map.has` in V8:
```cpp
Reduction JSCallReducer::ReduceMapPrototypeHas(Node* node) {
  return ReduceCollectionPrototypeHas(node, CollectionKind::kMap);
}

Reduction JSCallReducer::ReduceSetPrototypeHas(Node* node) {
  return ReduceCollectionPrototypeHas(node, CollectionKind::kSet);
}
```

I think for our case this "ignoreFields" config list is expected to be very small, so anyhow I see little value in harming instrumentation readability for this optimization.
Adding a `Map` from string to `true` just seems wired to me.

Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:67]: The config option is called "ignoreFields".

How about calling this field "_ignoreFieldsMap" (here and in other places)? Stating "ignore" without any context what is ignored is less readable IMO.

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:91]: nit: you can do it all in one go with no side effects:
```js
this._ignoreMap = new Map(vals?.map(fieldName => [fieldName, true]));
```

File: plugins/node/opentelemetry-instrumentation-graphql/test/graphql.test.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-graphql/test/graphql.test.ts:575]: Currently, instrumentation tests only with resolvers defined on the schema itself, but there are no tests for the patching of the "fieldResolver" (or defaultFieldResolver).

This PR adds logic that is currently not tested for the "fieldResolver".
Could be amazing to test your logic for this flow as well. Since there are no tests for this even for other features, I am ok if you choose not to do it.

File: plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts
Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts:67]: Originally it was using ".includes" so if you optimise something it is best to use the fastest way imho. This might be called thousands of times during usage so I would say it is still better to use less cpu than more. This is a kind of thing you do once and then you probably wont be touching this again for a long time. 
```
ignoreFields" config list is expected to be very small
```
what if not ?, besides that the list can be small but still it can be called many times per second.

**Comment by blumamir** [comment]
@jscherer92 Thanks for the time you put into this. I apologize for taking long to review.
Are you still interested in it or should I close?

**Comment by jscherer92** [comment]
Definitely still interested! I can take a look at the comments and fix up what is there

**Pull request closed without merging**
